### PR TITLE
[libc][NFC] Fix "type qualifiers ignored on cast result type"  GCC warning

### PR DESCRIPTION
**Pull request by gchatelet**
GCC complains about "type qualifiers ignored on cast result type"

### CODE REVIEW

**Comment by gchatelet** [comment]
Looking at functions around this one, I suspect the volatile is unnecessary.

**Comment by nickdesaulniers** [comment]
added by @lntue in d6802581700bd.

Honestly, does this test need to use `volatile` at all?  `volatile` is kind of a broken qualifier at least in regards to storage; it should only be used to force loads/stores to be atomic and non-coalesced (by casting the address of an lvalue to a pointer of that type plus volatile, then deferencing that).

**Comment by nickdesaulniers** [comment]
consider adding the diagnostic you're addressing to the commit message (here and in future patches) to help reviewers understand precisely what you're addressing.

**Comment by nickdesaulniers** [comment]
Also, I'm curious if clang supports this diagnostic flag and if/why we don't see it with clang. That's another "thread to pull on."

**Comment by lntue** [comment]
The intention of `volatile` was to force the conversion to double for each rounding mode inside the macro `EXPECT_FP_EQ_ALL_ROUNDING`: https://github.com/llvm/llvm-project/blob/main/libc/test/UnitTest/FPMatcher.h#L190
Without forcing re-evaluation, the compiler will optimize and only use the default rounding mode output of the conversion for all the rounding modes.
It seems to work properly on clang, and probably because gcc ignores the attribute, this test failed on gcc?

A proper fix is probably updating the `EXPECT_FP_EQ_ALL_ROUNDING` in every rounding mode check as:
```
  if (__r.success) {
    volatile auto e = (expected);
    volatile auto a = (actual);
    EXPECT_FP_EQ(e, a);
  }
```

**Comment by nickdesaulniers** [comment]
> this test failed on gcc?

Is this test failing with GCC?

**Comment by gchatelet** [comment]
Then a proper fix is probably to use something like google benchmark [`DoNotOptimize`](https://github.com/google/benchmark/blob/3d293cd67a264701378c46e2ae3b6408d533e093/include/benchmark/benchmark.h#L453-L456).

`volatile` is [broken in many ways](https://www.youtube.com/watch?v=KJW_DLaVXIY).

**Comment by nickdesaulniers** [comment]
> Without forcing re-evaluation, the compiler will optimize and only use the default rounding mode output of the conversion for all the rounding modes.

As in the compiler may do constant folding with a default rounding mode (rather than a rounding mode that may be changed during runtime)?

Regardless, could use comments in the source then to document _why_.

**Comment by gchatelet** [comment]
> > this test failed on gcc?
> 
> Is this test failing with GCC?

It doesn't AFAICT, GCC warns about "type qualifiers ignored on cast result type".

**Comment by nickdesaulniers** [comment]
> As in the compiler may do constant folding with a default rounding mode (rather than a rounding mode that may be changed during runtime)?

https://gcc.gnu.org/onlinedocs/gcc/Optimize-Options.html#index-frounding-math

**Comment by gchatelet** [comment]
I changed the build system to add `-frounding-math` to both CMake and Bazel. @lntue let me know if there's a better way to do this. I've checked that the flag is passed down to the command line but I haven't checked the generated code.

**Comment by lntue** [comment]
> I changed the build system to add `-frounding-math` to both CMake and Bazel. @lntue let me know if there's a better way to do this. I've checked that the flag is passed down to the command line but I haven't checked the generated code.

Adding that flag should work.  Can you only add that flag to the updated tests in this patch for now?

**Comment by gchatelet** [comment]
> > I changed the build system to add `-frounding-math` to both CMake and Bazel. @lntue let me know if there's a better way to do this. I've checked that the flag is passed down to the command line but I haven't checked the generated code.
> 
> Adding that flag should work. Can you only add that flag to the updated tests in this patch for now?

Do you mean that I should add the flag only on this particular test and not globally for all `add_fp_unittest` / `libc_test`?

**Comment by lntue** [comment]
> > > I changed the build system to add `-frounding-math` to both CMake and Bazel. @lntue let me know if there's a better way to do this. I've checked that the flag is passed down to the command line but I haven't checked the generated code.
> > 
> > 
> > Adding that flag should work. Can you only add that flag to the updated tests in this patch for now?
> 
> Do you mean that I should add the flag only on this particular test and not globally for all `add_fp_unittest` / `libc_test`?

Yes.

**Comment by nickdesaulniers** [comment]
> > Do you mean that I should add the flag only on this particular test and not globally for all `add_fp_unittest` / `libc_test`?
> 
> Yes.

+1, then you can remove all uses of `volatile` in that test as well.

**Review by nickdesaulniers — CHANGES_REQUESTED** [review]
don't forget to remove volatiles!

**Review by nickdesaulniers — APPROVED** [review]

File: libc/test/src/__support/FPUtil/CMakeLists.txt [hunk 1]
```diff
@@ -9,6 +9,8 @@ add_fp_unittest(
     dyadic_float_test.cpp
   DEPENDS
     libc.src.__support.FPUtil.dyadic_float
+  COMPILE_OPTIONS
```
Comment (nickdesaulniers) [inline libc/test/src/__support/FPUtil/CMakeLists.txt:12]: consider adding a comment why this is necessary

**Review by lntue — APPROVED** [review]

**Pull request merged**